Mu2eKinKal/src/LoopHelixFit_module.cc

```diff
@@ -347,88 +475,86 @@ namespace mu2e {
       nseed += hseedcol.size();
       // loop over the seeds
       for(size_t iseed=0; iseed < hseedcol.size(); ++iseed) {
+        ++nSeen_;
         auto const& hseed = hseedcol[iseed];
+        if(!goodHelix(hseed)) continue; //determine if this seed should be fit
         auto hptr = HPtr(hseedcol_h,iseed);
-        // check helicity.  The test on the charge and helicity
-        if(hseed.status().hasAllProperties(goodseed_) ){
-          // test helix
-          auto const& helix = hseed.helix();
-          if(helix.radius() == 0.0 || helix.lambda() == 0.0 )
-            throw cet::exception("RECO")<<"mu2e::HelixFit: degenerate seed parameters" << endl;
-          auto zcent = Mu2eKinKal::zMid(hseed.hits());
-          // take the magnetic field at the helix center as nominal
-          VEC3 center(helix.centerx(), helix.centery(),zcent);
-          auto bnom = kkbf_->fieldVect(center);
-          // compute the charge from the helicity, fit direction, and BField direction
-          double bz = bnom.Z();
-          int charge = static_cast<int>(copysign(PDGcharge_,(-1)*helix.helicity().value()*fdir_.dzdt()*bz));
-          // test consistency.  Modify this later when the HelixSeed knows which direction it's going TODO
-          auto fitpart = fpart_;
-          if(charge*PDGcharge_ < 0){
-            if(usePDGCharge_)throw cet::exception("RECO")<<"mu2e::HelixFit: inconsistent charge" << endl;
-            fitpart = static_cast<PDGCode::type>(-1*fitpart); // reverse sign
+
+        // determine the fit direction hypothesis
+        auto helix_dir = chooseHelixDir(hseed);
+        const bool undefined_dir = helix_dir == HelixRecoDir::PropDir::ambiguous; //check if no direction is significantly preferred
+
+        // fit the track hypothesis (hypotheses)
+        std::unique_ptr<KKTRK> ktrk = nullptr;
+        if(undefined_dir) {
+          ++nAmbiguous_;
+          auto ktrk_down = fitTrack(event, hseed,  1.f, fpart_);
+          auto ktrk_up   = fitTrack(event, hseed, -1.f, fpart_);
+          const bool use_downstream = !ktrk_up || (ktrk_down && compare_tracks(*ktrk_down, *ktrk_up)); //if one is null, default to the other (use down if both are null, doesn't matter)
+          if(print_ > 0)
+            printf("[LoopHelixFit::%s]: Fit both direction options, slope = %9.2e, sig = %.2f, status down = %i, status up = %i, p(chi^2) down = %.4f, p(chi^2) up = %.4f: chose %s\n",
+                   __func__, hseed.recoDir().slope(), hseed.recoDir().slopeSig(),
```

Comment:
This statement is too complex. Please recode it as a function.

>

Reply:
I've updated to pass the pointers rather than the references, so the `compare_tracks` function can also check if the pointer is null (though I'm not as used to working with `std::unique_ptr` so let me know if I haven't used it correctly)